packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,530 @@
+ * ║               |-------|-------|----║--|-------|----║
+ * ║                      ^        ^    ║ ^       ^     ║
+ * ║                                    ║  ^       ^    ║
+ * ╚════════════════════════════════════╝═══════════════╝
+ */
+export class PersistentCheckpointStateCache implements CheckpointStateCache {
+  private readonly cache: MapTracker<string, CachedBeaconStateAllForks | PersistentKey>;
+  // maintain order of epoch to decide which epoch to prune from memory
+  private readonly inMemoryEpochs: Set<Epoch>;
+  /** Epoch -> Set<blockRoot> */
+  private readonly epochIndex = new MapDef<Epoch, Set<string>>(() => new Set<string>());
+  private readonly metrics: Metrics["cpStateCache"] | null | undefined;
+  private readonly logger: Logger;
+  private readonly clock: IClock | null | undefined;
+  private preComputedCheckpoint: string | null = null;
+  private preComputedCheckpointHits: number | null = null;
+  private readonly maxEpochsInMemory: number;
+  private readonly persistentApis: CPStatePersistentApis;
+  private readonly shufflingCache: ShufflingCache;
+  private readonly getHeadState?: GetHeadStateFn;
+
+  constructor(
+    {metrics, logger, clock, shufflingCache, persistentApis, getHeadState}: PersistentCheckpointStateCacheModules,
+    opts: PersistentCheckpointStateCacheOpts
+  ) {
+    this.cache = new MapTracker(metrics?.cpStateCache);
+    if (metrics) {
+      this.metrics = metrics.cpStateCache;
+      metrics.cpStateCache.size.addCollect(() => {
+        let persistCount = 0;
+        let stateCount = 0;
+        const memoryEpochs = new Set<Epoch>();
+        const persistentEpochs = new Set<Epoch>();
+        for (const [key, stateOrPersistentKey] of this.cache.entries()) {
+          const {epoch} = fromCheckpointKey(key);
+          if (isPersistentKey(stateOrPersistentKey)) {
+            persistCount++;
+            memoryEpochs.add(epoch);
+          } else {
+            stateCount++;
+            persistentEpochs.add(epoch);
+          }
+        }
+        metrics.cpStateCache.size.set({type: CacheType.persistence}, persistCount);
+        metrics.cpStateCache.size.set({type: CacheType.state}, stateCount);
+        metrics.cpStateCache.epochSize.set({type: CacheType.persistence}, persistentEpochs.size);
+        metrics.cpStateCache.epochSize.set({type: CacheType.state}, memoryEpochs.size);
+      });
+    }
+    this.logger = logger;
+    this.clock = clock;
+    if (opts.maxEpochsInMemory < 0) {
+      throw new Error("maxEpochsInMemory must be >= 0");
+    }
+    this.maxEpochsInMemory = opts.maxEpochsInMemory;
+    // Specify different persistentApis for testing
+    this.persistentApis = persistentApis;
+    this.shufflingCache = shufflingCache;
+    this.getHeadState = getHeadState;
+    this.inMemoryEpochs = new Set();
+  }
+
+  /**
+   * Get a state from cache, it will reload from disk.
+   * This is expensive api, should only be called in some important flows:
```

Comment:
```suggestion
   * Get a state from cache, it may reload from disk.
   * This is an expensive api, should only be called in some important flows:
```